.config/emacs/config.org
#+TITLE: Emacs config
#+AUTHOR: Touhidul Shawan
#+DESCRIPTION: Shawan's Emacs configuration file
#+STARTUP: showeverything
#+OPTIONS: toc:2
* Table of Contents :toc:
- [[#package-manager][Package manager]]
- [[#use-package][use-package]]
- [[#editor][Editor]]
  - [[#defaults][defaults]]
  - [[#recentf][Recentf]]
  - [[#make-c-g-a-bit-helpful][Make C-g a bit helpful]]
  - [[#ui][UI]]
  - [[#font][Font]]
  - [[#theme][Theme]]
  - [[#line-number][Line number]]
  - [[#match-paren--auto-pair][match paren & auto pair]]
  - [[#backup--lock-file][backup & lock file]]
  - [[#minibuffer-escape][minibuffer escape]]
- [[#packages][Packages]]
  - [[#delsel][delsel]]
  - [[#evil-mode][evil mode]]
  - [[#nerd-icons][nerd icons]]
  - [[#vertico][vertico]]
  - [[#marginalia][marginalia]]
  - [[#orderless][orderless]]
  - [[#savehist][savehist]]
  - [[#corfu][corfu]]
  - [[#dired][dired]]
  - [[#dired-subtree][dired subtree]]
  - [[#trashed][trashed]]
  - [[#avy][avy]]
  - [[#consult][consult]]
  - [[#hl-todo][hl-todo]]
  - [[#modeline][modeline]]
  - [[#dashboard][dashboard]]
  - [[#yasnippet][yasnippet]]
- [[#org-mode][ORG mode]]
  - [[#agenda][Agenda]]
  - [[#capture][Capture]]
  - [[#toc][toc]]
  - [[#improve-visual][improve visual]]
  - [[#interaction][interaction]]
  - [[#org-tempo][org tempo]]
  - [[#org-modern][org modern]]
  - [[#keyword][keyword]]
- [[#keybindings][Keybindings]]
  - [[#which-key][which key]]
  - [[#evil][evil]]
  - [[#jj-to-normal-mode][jj to normal mode]]
  - [[#space-as-leader-key][space as leader key]]
  - [[#help][help]]
  - [[#files][files]]
  - [[#buffer][buffer]]
  - [[#window][window]]
  - [[#avy-1][avy]]
  - [[#magit][magit]]
  - [[#snippet][snippet]]
  - [[#org][org]]
  - [[#consult-1][consult]]

* Package manager
#+begin_src emacs-lisp
(require 'package)
(package-initialize)
  
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
#+end_src
* use-package
#+begin_src emacs-lisp
  (when (< emacs-major-version 29)
    (unless (package-installed-p 'use-package)
      (unless package-archive-contents
        (package-refresh-contents))
      (package-install 'use-package)))

  (add-to-list 'display-buffer-alist
             '("\\`\\*\\(Warnings\\|Compile-Log\\)\\*\\'"
               (display-buffer-no-window)
               (allow-no-window . t)))
#+end_src
* Editor
** defaults
#+begin_src emacs-lisp
  (delete-selection-mode 1)    ;; You can select text and delete it by typing.
  (electric-indent-mode -1)    ;; Turn off the weird indenting that Emacs does by default.
  (electric-pair-mode 1)       ;; Turns on automatic parens pairing
  ;; The following prevents <> from auto-pairing when electric-pair-mode is on.
  ;; Otherwise, org-tempo is broken when you try to <s TAB...
  (add-hook 'org-mode-hook (lambda ()
             (setq-local electric-pair-inhibit-predicate
                     `(lambda (c)
                    (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))
  (global-auto-revert-mode t)  ;; Automatically show changes if the file has changed
  (global-display-line-numbers-mode 1) ;; Display line numbers
  (global-visual-line-mode t)  ;; Enable truncated lines
  (setq org-edit-src-content-indentation 0) ;; Set src block automatic indent to 0 instead of 2.
  (setq use-file-dialog nil)   ;; No file dialog
  (setq use-dialog-box nil)    ;; No dialog box
  (setq pop-up-windows nil)    ;; No popup windows
#+end_src
** Recentf
#+begin_src emacs-lisp
(recentf-mode t)
#+end_src
** Make C-g a bit helpful
#+begin_src emacs-lisp
  (defun prot/keyboard-quit-dwim ()
  "Do-What-I-Mean behaviour for a general `keyboard-quit'.

The generic `keyboard-quit' does not do the expected thing when
the minibuffer is open.  Whereas we want it to close the
minibuffer, even without explicitly focusing it.

The DWIM behaviour of this command is as follows:

- When the region is active, disable it.
- When a minibuffer is open, but not focused, close the minibuffer.
- When the Completions buffer is selected, close it.
- In every other case use the regular `keyboard-quit'."
  (interactive)
  (cond
   ((region-active-p)
    (keyboard-quit))
   ((derived-mode-p 'completion-list-mode)
    (delete-completion-window))
   ((> (minibuffer-depth) 0)
    (abort-recursive-edit))
   (t
    (keyboard-quit))))

(define-key global-map (kbd "C-g") #'prot/keyboard-quit-dwim)

#+end_src
** UI
#+begin_src emacs-lisp
  (menu-bar-mode -1)
  (scroll-bar-mode -1)
  (tool-bar-mode -1)
  (setq inhibit-startup-message t)
  (setq use-short-answers t)
  (setq
   initial-major-mode 'org-mode
   initial-scratch-message ""
   initial-buffer-choice t) 
  (setq split-height-threshold nil)
  (setq split-width-threshold 0)
#+end_src
** Font
#+begin_src emacs-lisp
  (set-face-attribute 'default nil
                    :font "JetBrains Mono"
                    :height 111
                    :weight 'medium)
(set-face-attribute 'variable-pitch nil
                    :font "Ubuntu"
                    :height 122
                    :weight 'medium)
(set-face-attribute 'fixed-pitch nil
                    :font "JetBrains Mono"
                    :height 111
                    :weight 'medium)
(set-face-attribute 'font-lock-comment-face nil :slant 'italic)

(set-face-attribute 'font-lock-keyword-face nil :slant 'italic)

(add-to-list
 'default-frame-alist '(font . "JetBrains Mono-13"))
#+end_src
** Theme
#+begin_src emacs-lisp
(load-theme 'wombat t)
#+end_src
** Line number 
#+begin_src emacs-lisp
(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)
#+end_src
** match paren & auto pair
#+begin_src emacs-lisp
  (show-paren-mode 1)
  (electric-pair-mode 1)
#+end_src
** backup & lock file
disable backup and lock file
#+begin_src emacs-lisp
(setq make-backup-files nil) 
(setq create-lockfiles nil)  
#+end_src
** minibuffer escape 
#+begin_src emacs-lisp
(global-set-key [escape] 'keyboard-escape-quit)
#+end_src
* Packages
** delsel
- Delete the selected text upon text insertion
#+begin_src
(use-package delsel
  :ensure nil ; no need to install it as it is built-in
  :hook (after-init . delete-selection-mode))
#+end_src
** evil mode
*** evil
#+begin_src emacs-lisp
  (use-package evil
    :ensure t
    :demand t
    :bind (("<escape>" . keyboard-escape-quit))
    :init
    (setq
     evil-want-integration t
     evil-want-keybinding nil
     evil-vsplit-window-right t
     evil-split-window-below t
     evil-search-module 'evil-search
     evil-want-keybinding nil
     evil-disable-insert-state-bindings t
     evil-want-Y-yank-to-eol t
     evil-undo-system 'undo-redo)
    (evil-mode)
    :config (evil-set-leader 'normal " ") (evil-mode 1))
 #+end_src
*** evil collection
#+begin_src emacs-lisp
  (use-package evil-collection
  :ensure t
  :after evil
  :config
  (setq evil-want-integration t)
  (evil-collection-init))
#+end_src
*** evil surround  
#+begin_src emacs-lisp
(use-package evil-surround
  :ensure t
  :after evil
  :config (global-evil-surround-mode 1))
#+end_src
*** evil commentry 
#+begin_src emacs-lisp
  (use-package evil-commentary
    :ensure t
    :after evil
    :bind (:map evil-normal-state-map ("gc" . evil-commentary)))
#+end_src
*** evil org 
#+begin_src emacs-lisp
(use-package evil-org
  :ensure t
  :after org
  :config
  (require 'evil-org-agenda)
  (evil-org-agenda-set-keys)
  (add-hook 'org-mode-hook (lambda () (evil-org-mode 1))))
#+end_src
** nerd icons 
#+begin_src emacs-lisp
  (use-package nerd-icons
    :ensure t)
  (use-package nerd-icons-completion
    :ensure t
    :after marginalia
    :config
    (add-hook 'marginalia-mode-hook #'nerd-icons-completion-marginalia-setup))

  (use-package nerd-icons-corfu
    :ensure t
    :after corfu
    :config
    (add-to-list 'corfu-margin-formatters #'nerd-icons-corfu-formatter))
  (use-package nerd-icons-dired
    :ensure t
    :hook
    (dired-mode . nerd-icons-dired-mode))
#+end_src
** vertico
#+begin_src emacs-lisp
(use-package vertico
  :ensure t
  :hook (after-init . vertico-mode))
#+end_src
** marginalia 
#+begin_src  emacs-lisp
(use-package marginalia
  :ensure t
  :hook (after-init . marginalia-mode))
#+end_src
** orderless
#+begin_src emacs-lisp
(use-package orderless
  :ensure t
  :config
  (setq completion-styles '(orderless basic))
  (setq completion-category-defaults nil)
  (setq completion-category-overrides nil))
#+end_src
** savehist
#+begin_src emacs-lisp
(use-package savehist
  :ensure nil ; it is built-in
  :hook (after-init . savehist-mode))
#+end_src
** corfu
#+begin_src emacs-lisp
(use-package corfu
  :ensure t
  :hook (after-init . global-corfu-mode)
  :bind (:map corfu-map ("<tab>" . corfu-complete))
  :config
  (setq tab-always-indent 'complete)
  (setq corfu-preview-current nil)
  (setq corfu-min-width 20)

  (setq corfu-popupinfo-delay '(1.25 . 0.5))
  (corfu-popupinfo-mode 1) ; shows documentation after `corfu-popupinfo-delay'

  ;; Sort by input history (no need to modify `corfu-sort-function').
  (with-eval-after-load 'savehist
    (corfu-history-mode 1)
    (add-to-list 'savehist-additional-variables 'corfu-history)))
#+end_src
** dired
#+begin_src emacs-lisp
(use-package dired
  :ensure nil
  :commands (dired)
  :hook
  ((dired-mode . dired-hide-details-mode)
   (dired-mode . hl-line-mode))
  :config
  (setq dired-recursive-copies 'always)
  (setq dired-recursive-deletes 'always)
  (setq delete-by-moving-to-trash t)
  (setq dired-dwim-target t))
#+end_src
** dired subtree
#+begin_src emacs-lisp
(use-package dired-subtree
  :ensure t
  :after dired
  :bind
  ( :map dired-mode-map
    ("<tab>" . dired-subtree-toggle)
    ("TAB" . dired-subtree-toggle)
    ("<backtab>" . dired-subtree-remove)
    ("S-TAB" . dired-subtree-remove))
  :config
  (setq dired-subtree-use-backgrounds nil))
#+end_src
** trashed
#+begin_src emacs-lisp
(use-package trashed
  :ensure t
  :commands (trashed)
  :config
  (setq trashed-action-confirmer 'y-or-n-p)
  (setq trashed-use-header-line t)
  (setq trashed-sort-key '("Date deleted" . t))
  (setq trashed-date-format "%Y-%m-%d %H:%M:%S"))
#+end_src
** avy
#+begin_src emacs-lisp
  (use-package avy
    :ensure t
    :defer t
    :config
    (setq avy-case-fold-search nil))
#+end_src
** consult
#+begin_src emacs-lisp
(use-package consult :ensure t)
#+end_src
** hl-todo 
#+begin_src emacs-lisp
  (use-package hl-todo
    :ensure t
    :hook ((org-mode . hl-todo-mode)
           (prog-mode . hl-todo-mode))
    :config
    (setq hl-todo-highlight-punctuation ":"
          hl-todo-keyword-faces
          `(("TODO"       warning bold)
            ("FIXME"      error bold)
            ("LATER"      font-lock-constant-face bold)
            ("REVIEW"     font-lock-keyword-face bold)
            ("NOTE"       success bold)
            ("DEPRECATED" font-lock-doc-face bold))))
#+end_src
** modeline 
#+begin_src emacs-lisp
(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :config
  (setq doom-modeline-height 35      ;; sets modeline height
        doom-modeline-bar-width 5    ;; sets right bar width
        doom-modeline-persp-name t   ;; adds perspective name to modeline
        doom-modeline-persp-icon t)) ;; adds folder icon next to persp name
#+end_src
** dashboard 
#+begin_src emacs-lisp
  (use-package dashboard
    :ensure t 
    :init
    (setq initial-buffer-choice 'dashboard-open)
    (setq initial-scratch-message nil)
    (setq dashboard-set-heading-icons t)
    (setq dashboard-set-file-icons t)
    ;; (setq dashboard-startup-banner 'logo) 
    (setq dashboard-center-content nil) ;; set to 't' for centered content
    (setq dashboard-items '((recents . 5)))

    :custom 
    (dashboard-modify-heading-icons '((recents . "file-text")))

    :config
    (dashboard-setup-startup-hook))
#+end_src
** yasnippet
#+begin_src emacs-lisp
  (use-package yasnippet
    :ensure t
    :config
    (setq yas-snippet-dirs '("~/.config/emacs/snippets"))
    (yas-global-mode 1))

#+end_src
* ORG mode
#+begin_src emacs-lisp
  (use-package org
    :ensure nil
    :config
    (setq org-insert-heading-respect-content t)
    (setq org-log-done 'time)
    (setq org-log-into-drawer t)
    (setq org-directory "~/shawan-org/"))
#+end_src
** Agenda
#+begin_src emacs-lisp
(setq org-agenda-files 
      (list (concat org-directory "tasks.org")
            (concat org-directory "journal.org")))
(setq org-agenda-window-setup 'only-window) 
#+end_src
** Capture
from DT
#+begin_src emacs-lisp
  ;; Defines the global fallback destination for all your Org notes.
  (setq org-default-notes-file (concat org-directory "notes.org"))

  ;; (Optional) Create custom templates
  ;; NOTE '%U' is an inactive timestamp meaning the item will not show
  ;; in org-agenda.  Use '%^t' for active timestamps instead.
  ;; You can manually switch active/inactive with SHIFT-up/down.
  (setq org-capture-templates
        `(
          ;; Idea capture
          ("i" "idea" entry
           (file ,org-default-notes-file)
           "* %? :idea: \n%U\n")

          ;; Journal entry
          ("j" "journal" entry
           (file+olp+datetree ,(concat org-directory "journal.org"))
           "* %U\n%?\n")

          ;; Note with link to source
          ("n" "note" entry
           (file ,org-default-notes-file)
  	 "* %? :note: \n%U\n")
  	
          ;; Todo with context
          ("t" "task" entry
           ;; (file+olp+datetree ,(concat org-directory "tasks.org"))
  	 (file ,(concat org-directory "tasks.org"))
           "* %?")

          ))
#+end_src
** toc
#+begin_src emacs-lisp
  (use-package toc-org
      :ensure t
      :commands toc-org-enable
      :init (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src
** improve visual
#+begin_src emacs-lisp
(setq org-ellipsis " ▾")
(setq org-src-fontify-natively t)
(setq org-highlight-latex-and-related '(native))
(setq org-startup-folded 'showeverything)
(setq org-startup-with-inline-images t)
(setq org-image-actual-width 300)
(setq org-fontify-whole-heading-line t)
(setq org-pretty-entities t)
(setq org-hide-emphasis-markers t)
(setq org-adapt-indentation t)
(setq org-startup-indented t)
(setq org-special-ctrl-a/e '(t . nil))
(setq org-special-ctrl-k t)
(setq org-fontify-quote-and-verse-blocks t)
(setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 2)
(setq org-hide-block-startup nil)
(setq org-src-preserve-indentation nil)
(setq org-startup-folded 'fold)
(setq org-cycle-separator-lines 2)
(setq org-goto-auto-isearch nil)
(setq org-log-done 'time)
(setq org-log-into-drawer t)
#+end_src
** interaction 
#+begin_src emacs-lisp
(setq org-cycle-separator-lines 1)
(setq org-catch-invisible-edits 'show-and-error)
(setq org-src-tab-acts-natively t)
#+end_src
** org tempo
#+begin_src emacs-lisp
(require 'org-tempo) 
#+end_src
** org modern 
#+begin_src emacs-lisp
(use-package org-modern
  :ensure t
  :hook (org-mode . org-modern-mode)
  :config
  (setq
   ;; org-modern-star '("●" "○" "✸" "✿")
   org-modern-star '("⌾" "✸" "◈" "✿")
   org-modern-list '((42 . "◦") (43 . "•") (45 . "–"))
   org-modern-tag nil
   org-modern-priority nil
   org-modern-todo nil
   org-modern-table nil
   org-modern-variable-pitch nil

.config/emacs/config.el
(require 'package)
(package-initialize)
  
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))

(when (< emacs-major-version 29)
  (unless (package-installed-p 'use-package)
    (unless package-archive-contents
      (package-refresh-contents))
    (package-install 'use-package)))

(add-to-list 'display-buffer-alist
           '("\\`\\*\\(Warnings\\|Compile-Log\\)\\*\\'"
             (display-buffer-no-window)
             (allow-no-window . t)))

(delete-selection-mode 1)    ;; You can select text and delete it by typing.
(electric-indent-mode -1)    ;; Turn off the weird indenting that Emacs does by default.
(electric-pair-mode 1)       ;; Turns on automatic parens pairing
;; The following prevents <> from auto-pairing when electric-pair-mode is on.
;; Otherwise, org-tempo is broken when you try to <s TAB...
(add-hook 'org-mode-hook (lambda ()
           (setq-local electric-pair-inhibit-predicate
                   `(lambda (c)
                  (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))
(global-auto-revert-mode t)  ;; Automatically show changes if the file has changed
(global-display-line-numbers-mode 1) ;; Display line numbers
(global-visual-line-mode t)  ;; Enable truncated lines
(setq org-edit-src-content-indentation 0) ;; Set src block automatic indent to 0 instead of 2.
(setq use-file-dialog nil)   ;; No file dialog
(setq use-dialog-box nil)    ;; No dialog box
(setq pop-up-windows nil)    ;; No popup windows

(recentf-mode t)

(defun prot/keyboard-quit-dwim ()
  "Do-What-I-Mean behaviour for a general `keyboard-quit'.

The generic `keyboard-quit' does not do the expected thing when
the minibuffer is open.  Whereas we want it to close the
minibuffer, even without explicitly focusing it.

The DWIM behaviour of this command is as follows:

- When the region is active, disable it.
- When a minibuffer is open, but not focused, close the minibuffer.
- When the Completions buffer is selected, close it.
- In every other case use the regular `keyboard-quit'."
  (interactive)
  (cond
   ((region-active-p)
    (keyboard-quit))
   ((derived-mode-p 'completion-list-mode)
    (delete-completion-window))
   ((> (minibuffer-depth) 0)
    (abort-recursive-edit))
   (t
    (keyboard-quit))))

(define-key global-map (kbd "C-g") #'prot/keyboard-quit-dwim)

(menu-bar-mode -1)
(scroll-bar-mode -1)
(tool-bar-mode -1)
(setq inhibit-startup-message t)
(setq use-short-answers t)
(setq
 initial-major-mode 'org-mode
 initial-scratch-message ""
 initial-buffer-choice t) 
(setq split-height-threshold nil)
(setq split-width-threshold 0)

(set-face-attribute 'default nil
                    :font "JetBrains Mono"
                    :height 111
                    :weight 'medium)
(set-face-attribute 'variable-pitch nil
                    :font "Ubuntu"
                    :height 122
                    :weight 'medium)
(set-face-attribute 'fixed-pitch nil
                    :font "JetBrains Mono"
                    :height 111
                    :weight 'medium)
(set-face-attribute 'font-lock-comment-face nil :slant 'italic)

(set-face-attribute 'font-lock-keyword-face nil :slant 'italic)

(add-to-list
 'default-frame-alist '(font . "JetBrains Mono-13"))

(load-theme 'wombat t)

(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)

(show-paren-mode 1)
(electric-pair-mode 1)

(setq make-backup-files nil) 
(setq create-lockfiles nil)

(global-set-key [escape] 'keyboard-escape-quit)

(use-package evil
  :ensure t
  :demand t
  :bind (("<escape>" . keyboard-escape-quit))
  :init
  (setq
   evil-want-integration t
   evil-want-keybinding nil
   evil-vsplit-window-right t
   evil-split-window-below t
   evil-search-module 'evil-search
   evil-want-keybinding nil
   evil-disable-insert-state-bindings t
   evil-want-Y-yank-to-eol t
   evil-undo-system 'undo-redo)
  (evil-mode)
  :config (evil-set-leader 'normal " ") (evil-mode 1))

(use-package evil-collection
:ensure t
:after evil
:config
(setq evil-want-integration t)
(evil-collection-init))

(use-package evil-surround
  :ensure t
  :after evil
  :config (global-evil-surround-mode 1))

(use-package evil-commentary
  :ensure t
  :after evil
  :bind (:map evil-normal-state-map ("gc" . evil-commentary)))

(use-package evil-org
  :ensure t
  :after org
  :config
  (require 'evil-org-agenda)
  (evil-org-agenda-set-keys)
  (add-hook 'org-mode-hook (lambda () (evil-org-mode 1))))

(use-package nerd-icons
  :ensure t)
(use-package nerd-icons-completion
  :ensure t
  :after marginalia
  :config
  (add-hook 'marginalia-mode-hook #'nerd-icons-completion-marginalia-setup))

(use-package nerd-icons-corfu
  :ensure t
  :after corfu
  :config
  (add-to-list 'corfu-margin-formatters #'nerd-icons-corfu-formatter))
(use-package nerd-icons-dired
  :ensure t
  :hook
  (dired-mode . nerd-icons-dired-mode))

(use-package vertico
  :ensure t
  :hook (after-init . vertico-mode))

(use-package marginalia
  :ensure t
  :hook (after-init . marginalia-mode))

(use-package orderless
  :ensure t
  :config
  (setq completion-styles '(orderless basic))
  (setq completion-category-defaults nil)
  (setq completion-category-overrides nil))

(use-package savehist
  :ensure nil ; it is built-in
  :hook (after-init . savehist-mode))

(use-package corfu
  :ensure t
  :hook (after-init . global-corfu-mode)
  :bind (:map corfu-map ("<tab>" . corfu-complete))
  :config
  (setq tab-always-indent 'complete)
  (setq corfu-preview-current nil)
  (setq corfu-min-width 20)

  (setq corfu-popupinfo-delay '(1.25 . 0.5))
  (corfu-popupinfo-mode 1) ; shows documentation after `corfu-popupinfo-delay'

  ;; Sort by input history (no need to modify `corfu-sort-function').
  (with-eval-after-load 'savehist
    (corfu-history-mode 1)
    (add-to-list 'savehist-additional-variables 'corfu-history)))

(use-package dired
  :ensure nil
  :commands (dired)
  :hook
  ((dired-mode . dired-hide-details-mode)
   (dired-mode . hl-line-mode))
  :config
  (setq dired-recursive-copies 'always)
  (setq dired-recursive-deletes 'always)
  (setq delete-by-moving-to-trash t)
  (setq dired-dwim-target t))

(use-package dired-subtree
  :ensure t
  :after dired
  :bind
  ( :map dired-mode-map
    ("<tab>" . dired-subtree-toggle)
    ("TAB" . dired-subtree-toggle)
    ("<backtab>" . dired-subtree-remove)
    ("S-TAB" . dired-subtree-remove))
  :config
  (setq dired-subtree-use-backgrounds nil))

(use-package trashed
  :ensure t
  :commands (trashed)
  :config
  (setq trashed-action-confirmer 'y-or-n-p)
  (setq trashed-use-header-line t)
  (setq trashed-sort-key '("Date deleted" . t))
  (setq trashed-date-format "%Y-%m-%d %H:%M:%S"))

(use-package avy
  :ensure t
  :defer t
  :config
  (setq avy-case-fold-search nil))

(use-package consult :ensure t)

(use-package hl-todo
  :ensure t
  :hook ((org-mode . hl-todo-mode)
         (prog-mode . hl-todo-mode))
  :config
  (setq hl-todo-highlight-punctuation ":"
        hl-todo-keyword-faces
        `(("TODO"       warning bold)
          ("FIXME"      error bold)
          ("LATER"      font-lock-constant-face bold)
          ("REVIEW"     font-lock-keyword-face bold)
          ("NOTE"       success bold)
          ("DEPRECATED" font-lock-doc-face bold))))

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :config
  (setq doom-modeline-height 35      ;; sets modeline height
        doom-modeline-bar-width 5    ;; sets right bar width
        doom-modeline-persp-name t   ;; adds perspective name to modeline
        doom-modeline-persp-icon t)) ;; adds folder icon next to persp name

(use-package dashboard
  :ensure t 
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq initial-scratch-message nil)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  ;; (setq dashboard-startup-banner 'logo) 
  (setq dashboard-center-content nil) ;; set to 't' for centered content
  (setq dashboard-items '((recents . 5)))

  :custom 
  (dashboard-modify-heading-icons '((recents . "file-text")))

  :config
  (dashboard-setup-startup-hook))

(use-package yasnippet
  :ensure t
  :config
  (setq yas-snippet-dirs '("~/.config/emacs/snippets"))
  (yas-global-mode 1))

(use-package org
  :ensure nil
  :config
  (setq org-insert-heading-respect-content t)
  (setq org-log-done 'time)
  (setq org-log-into-drawer t)
  (setq org-directory "~/shawan-org/"))

(setq org-agenda-files 
      (list (concat org-directory "tasks.org")
            (concat org-directory "journal.org")))
(setq org-agenda-window-setup 'only-window)

;; Defines the global fallback destination for all your Org notes.
(setq org-default-notes-file (concat org-directory "notes.org"))

;; (Optional) Create custom templates
;; NOTE '%U' is an inactive timestamp meaning the item will not show
;; in org-agenda.  Use '%^t' for active timestamps instead.
;; You can manually switch active/inactive with SHIFT-up/down.
(setq org-capture-templates
      `(
        ;; Idea capture
        ("i" "idea" entry
         (file ,org-default-notes-file)
         "* %? :idea: \n%U\n")

        ;; Journal entry
        ("j" "journal" entry
         (file+olp+datetree ,(concat org-directory "journal.org"))
         "* %U\n%?\n")

        ;; Note with link to source
        ("n" "note" entry
         (file ,org-default-notes-file)
	 "* %? :note: \n%U\n")
	
        ;; Todo with context
        ("t" "task" entry
         ;; (file+olp+datetree ,(concat org-directory "tasks.org"))
	 (file ,(concat org-directory "tasks.org"))
         "* %?")

        ))

(use-package toc-org
    :ensure t
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))

(setq org-ellipsis " ▾")
(setq org-src-fontify-natively t)
(setq org-highlight-latex-and-related '(native))
(setq org-startup-folded 'showeverything)
(setq org-startup-with-inline-images t)
(setq org-image-actual-width 300)
(setq org-fontify-whole-heading-line t)
(setq org-pretty-entities t)
(setq org-hide-emphasis-markers t)
(setq org-adapt-indentation t)
(setq org-startup-indented t)
(setq org-special-ctrl-a/e '(t . nil))
(setq org-special-ctrl-k t)
(setq org-fontify-quote-and-verse-blocks t)
(setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 2)
(setq org-hide-block-startup nil)
(setq org-src-preserve-indentation nil)
(setq org-startup-folded 'fold)
(setq org-cycle-separator-lines 2)
(setq org-goto-auto-isearch nil)
(setq org-log-done 'time)
(setq org-log-into-drawer t)

(setq org-cycle-separator-lines 1)
(setq org-catch-invisible-edits 'show-and-error)
(setq org-src-tab-acts-natively t)

(require 'org-tempo)

(use-package org-modern
  :ensure t
  :hook (org-mode . org-modern-mode)
  :config
  (setq
   ;; org-modern-star '("●" "○" "✸" "✿")
   org-modern-star '("⌾" "✸" "◈" "✿")
   org-modern-list '((42 . "◦") (43 . "•") (45 . "–"))
   org-modern-tag nil
   org-modern-priority nil
   org-modern-todo nil
   org-modern-table nil
   org-modern-variable-pitch nil
   org-modern-block-fringe nil))

(setq org-todo-keywords
      '((sequence "TODO(t!)" "FAILED(f!)" "|" "DONE(d!)")
        (sequence
         "DROP(o!)"
         "WORK-IN-PROGRESS(w!)"
         "POSTPONE(p!)")))

(setq org-todo-keyword-faces
      '(("TODO"
         :inherit (region org-todo)
         :foreground "DarkOrange1"
         :weight bold)
        ("FAILED"
         :inherit (region org-todo)
         :foreground "white smoke"
         :background "dark red"
         :weight bold)
        ("WIP"
         :inherit (org-todo region)
         :foreground "magenta3"
         :weight bold)
        ("DONE" . "SeaGreen4")))

(use-package which-key
:ensure t
:init (which-key-mode 1)
:config
(setq
 which-key-side-window-location 'bottom
 which-key-sort-order #'which-key-key-order-alpha
 which-key-sort-uppercase-first nil
 which-key-add-column-padding 1
 which-key-max-display-columns nil
 which-key-min-display-lines 6
 which-key-side-window-slot -10
 which-key-side-window-max-height 0.25
 which-key-idle-delay 0.8
 which-key-max-description-length 25
 which-key-allow-imprecise-window-fit t
 which-key-prefix-prefix "◉ "
 which-key-separator " → "))

(use-package general
  :ensure t
  :config (general-evil-setup)

(general-imap
  "j" (general-key-dispatch 'self-insert-command
        :timeout 0.2 "j" 'evil-normal-state))

(general-create-definer
leader-key
:states '(normal insert visual emacs)
:keymaps 'override
:prefix "SPC"
:global-prefix "M-SPC")

(leader-key
  "h" '(:ignore t :wk "Help")
  "h f" '(describe-function :wk "Describe function")
  "h v" '(describe-variable :wk "Describe variable")
  "h r r" '((lambda ()
              (interactive) (load-file "~/.config/emacs/init.el"))
            :wk "Reload emacs config"))

(leader-key 
  "f" '(:ignore t :wk "Files")
  "." '(find-file :wk "Find file")
  "f f" '(find-file :wk "Find file")
  "f c" '((lambda ()
            (interactive)
            (find-file "~/.config/emacs/config.org"))
          :wk "Edit emacs config")
  "f s" '(save-buffer :wk "Save buffer")
  "f r" '(consult-recent-file :wk "Find recent files")
  "f q" '(kill-buffer :wk "Kill buffer"))

(leader-key
  "b" '(:ignore t :wk "buffer")
  "b i" '(ibuffer :wk "Switch ibuffer")
  "b b" '(switch-to-buffer :wk "Switch buffer")
  "b k" '(kill-current-buffer :wk "Kill this buffer")
  "b n" '(next-buffer :wk "Next buffer")
  "b p" '(previous-buffer :wk "Previous buffer")
  "b r" '(revert-buffer :wk "Reload buffer")
  "b s" '(scratch-buffer :wk "Scratch buffer"))

(leader-key
  "w" '(:ignore t :wk "Windows")
  "w c" '(evil-window-delete :wk "Close window")
  "w n" '(evil-window-new :wk "New window")
  "w s" '(evil-window-split :wk "Horizontal split window")
  "w v" '(evil-window-vsplit :wk "Vertical split window")
  "w h" '(evil-window-left :wk "Window left")
  "w j" '(evil-window-down :wk "Window down")
  "w k" '(evil-window-up :wk "Window up")
  "w l" '(evil-window-right :wk "Window right")
  "w w" '(evil-window-next :wk "Goto next window")
  "w H" '(buf-move-left :wk "Buffer move left")
  "w J" '(buf-move-down :wk "Buffer move down")
  "w K" '(buf-move-up :wk "Buffer move up")
  "w L" '(buf-move-right :wk "Buffer move right"))

(leader-key
  "j" '(avy-goto-word-0 :wk "Go to word")
  "l" '(avy-goto-line :wk "Go to line"))

(leader-key
  "g" '(:ignore t :wk "magit")
  "g g" '(magit-status :wk "Magit Status"))

(leader-key
  "i" '(:ignore t :wk "snippets")
  "s" '(yas-insert-snippet :wk "Yas insert snippet"))

(leader-key
  "m" '(:ignore t :wk "Org")
  "m a" '(org-agenda :wk "Org agenda")
  "m c" '(org-capture :wk "Org capture")
  "m e" '(org-export-dispatch :wk "Org export dispatch")
  "m i" '(org-toggle-item :wk "Org toggle item")
  "m t" '(org-todo :wk "Org todo")
  "m B" '(org-babel-tangle :wk "Org babel tangle")
  "m T" '(org-todo-list :wk "Org todo list"))
(leader-key
  "m b" '(:ignore t :wk "Tables")
  "m b -" '(org-table-insert-hline :wk "Insert hline in table"))

(leader-key
  "m d" '(:ignore t :wk "Date/deadline")
  "m d t" '(org-time-stamp :wk "Org time stamp"))

(leader-key
  "m i" '(org-toggle-inline-images :wk "Toggle inline image"))

(leader-key
  "x" '(:ignore t :wk "Consult")
  "x b" '(consult-buffer :wk "consult buffer")
  "x y" '(consult-yank-pop :wk "consult yank pop")
  "x l" '(consult-goto-line :wk "consult goto-line")
  "x f" '(consult-flymake :wk "consult flymake")
  "x i" '(consult-imenu :wk "consult imenu")
  "x g" '(consult-ripgrep :wk "consult ripgre")
  "x x" '(consult-fd :wk "consult find")
  ))
